ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/fastsync/FastSyncDownloader.java

```diff
@@ -80,17 +84,21 @@ public FastSyncDownloader(
             .thenApply(this::updateMaxTrailingPeers)
             .thenApply(this::storeState)
             .thenCompose(this::downloadChainAndWorldState),
-        this::handleWorldStateUnavailable);
+        this::handleFailure);
   }
 
-  private CompletableFuture<FastSyncState> handleWorldStateUnavailable(final Throwable error) {
+  private CompletableFuture<FastSyncState> handleFailure(final Throwable error) {
     trailingPeerRequirements = Optional.empty();
-    if (ExceptionUtils.rootCause(error) instanceof StalledDownloadException) {
+    if (ExceptionUtils.rootCause(error) instanceof FastSyncException) {
+      return completedExceptionally(error);
+    } else if (ExceptionUtils.rootCause(error) instanceof StalledDownloadException) {
       LOG.warn(
           "Fast sync was unable to download the world state. Retrying with a new pivot block.");
       return start(FastSyncState.EMPTY_SYNC_STATE);
     } else {
-      return completedExceptionally(error);
+      LOG.error("Encountered an unexpected error during fast sync. Restarting fast sync.");
```

Comment:
We should log the error since it's we don't really want generic errors bubbling all the way up here.  Also, probably worth mentioning the delay here:
```suggestion
      LOG.error("Encountered an unexpected error during fast sync. Restarting fast sync in " + FAST_SYNC_RETRY_DELAY + " seconds.", error);
```

**Note**: To get stack traces printed in the logs, you can only supply a single error parameter to the `error` method.